### DISCUSSION TITLE
Import external model using `PyMechanical`

### DISCUSSION THREAD

**Discussion by narnia-ai-mason** [ansys/pymechanical#424] [Q&A] [answered]
Hi, I have asked about accessing Mechanical remotely the other day ([here](https://github.com/ansys/pymechanical/issues/417#issue-1912611975)).

The issue has been solved by @samigithub2022 which was running `AnsysWBU.exe` before trying `launch_mechanical(ip=my_ip)`:
```PowerShell
"C:\Program Files\ANSYS Inc\v231\aisol\bin\winx64\AnsysWBU.exe" -DSApplet -nosplash -notabctrl -grpc 10000
```

However, when I try to import external model by either GUI or Script `ExtAPI.Application.ImportMesh(model_path)` the following error occur:
![image](https://github.com/ansys/pymechanical/assets/144668347/04297ab2-9497-456b-9c31-bedc20748215)
(or a message "Mechanical locked when importing External Model.")

If I run Mechanical manually (or by `"C:\Program Files\ANSYS Inc\v231\Framework\bin\Win64\RunWB2.exe" -I -E "Mechanical.CreateMechanicalModelSystemAndOpen()"`), importing an external model works well.

Is it possible to import external model by `PyMechanical` and the corresponding Mechanical scripting?

**Comment by koubaa** [comment, marked as answer]
@narnia-mason We're considering supporting external model without workbench in future releases. In 2023R1 and 2023R2, it is not yet supported. I can't comment on what is coming in future releases until they are shipped.

What is your overall workflow? Import which file formats through external model into mechanical and then do what with it inside Mechanical?

**Reply by narnia-ai-mason** [reply]
Thank you for the comment!

This is for the case when I have the mesh information only not the corresponding geometry.
Also I'd like to control material data without engineering data in the Workbench.
The mesh can be delivered by any type; for example, `*.nas` bulkdata format.

In this case, can `pymapdl` be an alternative?

**Reply by koubaa** [reply]
@narnia-mason There is the project pymaterials-manager in pyansys to generate material model files in a format that is compatible wtih Mechanical.  It's still under construction, but that library may one day help you do what you need.

Thanks for sharing the file format. After that - I'm still not sure what you were originally looking for from PyMechanical. Suppose you can import that mesh into mechanical, what next? Would you have wanted to use Mechanical to solve using the MAPDL solver, or to generate a report, or to add boundary conditions and solve with a different solver, or something else?

**Comment by EntPyle** [comment]
plus one on this request for external model support. I'd like to be able to bring in surface meshes (any of obj, ply, stl would work)

My use case is I have statistical shape models of a surface mesh and I'm attempting to map that correspondence to paired volumetric data.

**Reply by pmaroneh** [reply]
@EntPyle Starting from 2024R1 it is possible to import an external mesh. However the only supported format is .cdb (Mechanical APDL common database). 
If you have .stl format (so, faceted geometry and not a mesh per say), one way to go could be to use PyAnsysGeometry to connect to a Discovery/SpaceClaim service, rebuild a geometry from the .stl file, then send that geometry to PyMechanical. Pinging @RobPasMue to confirm.

**Reply by RobPasMue** [reply]
HI @pmaroneh @EntPyle thanks for reaching out!

We support the generation of various formats using PyAnsys Geometry. You can find them here: https://geometry.docs.pyansys.com/version/stable/api/ansys/geometry/core/designer/design/DesignFileFormat.html#DesignFileFormat

**Reply by RobPasMue** [reply]
These formats are supported both in import and export, but starting on 24R1 and onwards

**Reply by RobPasMue** [reply]
Also, if you are using SpaceClaim directly you could try to import directly your STL file using the ``modeler.open_file()`` API

[Modeler.open_file — PyAnsys Geometry](https://geometry.docs.pyansys.com/version/stable/api/ansys/geometry/core/modeler/Modeler.html#Modeler.open_file)
